### PR TITLE
secondary_index: fix returned rows token ordering

### LINKED ISSUE #7443: idx_token of global secondary indexes is sorted incorrectly
The first column of clustering key in global secondary indexes is a token of partition key (of base table). `idx_token` is of type `bytes`. However tokens on ring are represented as **signed** `int64_t`. It is a problem, because rows are sorted by `idx_token` - bytes, but the ordering of bytes is **unsigned** (-1 is larger than 1 when doing unsigned comparisons):

```
cqlsh> SELECT pk, ck, ck2, token(pk) FROM ks.t;

 pk | ck | ck2 | system.token(pk)
----+----+-----+----------------------
  1 |  1 |   1 | -4069959284402364209
  1 |  1 |   5 | -4069959284402364209
  2 |  1 |   3 | -3248873570005575792
  2 |  1 |   8 | -3248873570005575792
  4 |  1 |   5 | -2729420104000364805
  3 |  1 |   5 |  9010454139840013625

cqlsh> SELECT pk, ck, ck2, token(pk) FROM ks.t WHERE ck = 1;

 pk | ck | ck2 | system.token(pk)
----+----+-----+----------------------
  3 |  1 |   5 |  9010454139840013625
  1 |  1 |   1 | -4069959284402364209
  1 |  1 |   5 | -4069959284402364209
  2 |  1 |   3 | -3248873570005575792
  2 |  1 |   8 | -3248873570005575792
  4 |  1 |   5 | -2729420104000364805

cqlsh> select * from ks.t_ck_idx_index;

 ck | idx_token          | pk | ck2
----+--------------------+----+-----
  1 | 0x7d0b904cfe1ca539 |  3 |   5
  1 | 0xc78499982ae4e0cf |  1 |   1
  1 | 0xc78499982ae4e0cf |  1 |   5
  1 | 0xd2e9aeb38ba99790 |  2 |   3
  1 | 0xd2e9aeb38ba99790 |  2 |   8
  1 | 0xda1f26e0623f9efb |  4 |   5
```

When doing `SELECT` on the base table without any filtering (it is not using secondary index in that case), `token(pk)` is correctly "sorted". However looking at `SELECT` with filtering (using secondary index), the order is not correct (it is ordered according to unsigned comparison). The same fact is true for the index view.

### PR DESCRIPTION
**Pull request by avelanarius**
Fixes returned rows ordering to proper signed token ordering. Before this change, rows were sorted by token, but using unsigned comparison, meaning that negative tokens appeared after positive tokens. 

Rename `token_column_computation` to `legacy_token_column_computation` and add some comments describing this computation.

Added (new) `token_column_computation` which returns token as `long_type`, which is sorted using signed comparison - the correct ordering of tokens.

Add new `correct_idx_token_in_secondary_index` feature, which flags that the whole cluster is able to use new `token_column_computation`.

Switch token computation in secondary indexes to (new) `token_column_computation`, which fixes the ordering. This column computation type is only set if cluster supports `correct_idx_token_in_secondary_index` feature to make sure that all nodes
will be able to compute new `token_column_computation`. Also old indexes will need to be rebuilt to take advantage of this fix, as new token column computation type is only set for new indexes.

Fix tests according to new token ordering and add one new test to validate this aspect explicitly.

Fixes #7443

Tested manually a scenario when someone created an index on old version of Scylla and then migrated to new Scylla. Old index continued to work properly (but returning in wrong order). Upon dropping and re-creating the index, it still returned the same data, but now in correct order.

### CODE REVIEW

File: bytes.hh
Comment (psarna) [inline bytes.hh:72]: This function makes sense only for integer types, right? A concept would be in order then (std::is_integral?).

File: column_computation.hh
Comment (psarna) [inline column_computation.hh:66]: Hm, maybe we should consider renaming the old computation to `legacy_token_column_computation` and name the new one `token_column_computation`. That way, nobody would get tempted to ever use the old, incorrect implementation by mistake. Or, you can simply wrap the old one in `legacy` namespace, which is customary.

File: index/secondary_index_manager.cc
Comment (psarna) [inline index/secondary_index_manager.cc:154]: That's what I was talking about above. This `token_column_computation` name sounds correct, so users might get tempted to use it instead of `unsigned_token_column_computation`, which sounds weird. Let's rename the correct version to `token_column_computation` and invent an ugly, deterring name for the old version - e.g. `legacy::token_column_computation`.

File: schema.cc
Comment (psarna) [inline schema.cc:1607]: :( That one has to stay, unfortunately.

**Review by psarna — COMMENTED** [review]
Looks good, left some ideological comments.

File: index/secondary_index_manager.cc
Comment (haaawk) [inline index/secondary_index_manager.cc:149]: GLOBAL ALERT!!!
Maybe store a reference to storage_service in secondary_index_manager as a field?

File: dht/token.hh
Comment (haaawk) [inline dht/token.hh:106]: This comment is a bit confusing. Can't we just add (1ull << 63) to unsigned_value?

**Comment by nyh** [comment]
> Fixes returned rows ordering to proper signed token ordering. Before this change, rows were sorted by token, but using unsigned comparison, meaning that negative tokens appeared after positive tokens.
> 
> Add new `unsigned_data()` method to `token`, which returns `token` as `bytes`, but first mapping `token` range from `[-2^63, 2^63-1]` to `[0, 2^64-1]`, which allows it to be compared using unsigned comparsions (such as `bytes` comparison) while preserving the correct signed token ordering.

What's the benefit of this approach over just storing the token as a signed integer? If I understand correctly, we already decided that the token will always be a 64-bit integer, and we'll never go back to other possibilities like bytes?

> 
> Add new `unsigned_token_column_computation`, which computes using new `unsigned_data()` method.
> 
> Add new `correct_idx_token_in_secondary_index` feature, which flags that the whole cluster is able to use new `unsigned_token_column_computation`.

We don't have to do this now, but I wonder if instead of adding a new cluster feature for every new computed-column computation, we could somehow make every computed column computation into a cluster feature? @psarna could that make sense? I'm asking this because we'll probably need such a feature for every computed-column computation class we want to add in the future.

> Fix tests according to new token ordering

Sad (it means we had tests which were wrong, and enshrined wrong behavior).

> and add one new test to validate this aspect explicitly.

Good. I also wrote a test for this feature in Python which I attached in the issue, which unfortunately I can't get into the repository because @kostja hates shell scripts and I didn't yet have time to convert it to Python.
 
> Fixes #7443
> 
> Tested manually a scenario when someone created an index on old version of Scylla and then migrated to new Scylla. Old index continued to work properly (but returning in wrong order). Upon dropping and re-creating the index, it still returned the same data, but now in correct order.

Great.

**Review by haaawk — APPROVED** [review]
LGTM

**Comment by haaawk** [comment]
> > Fixes returned rows ordering to proper signed token ordering. Before this change, rows were sorted by token, but using unsigned comparison, meaning that negative tokens appeared after positive tokens.
> > Add new `unsigned_data()` method to `token`, which returns `token` as `bytes`, but first mapping `token` range from `[-2^63, 2^63-1]` to `[0, 2^64-1]`, which allows it to be compared using unsigned comparsions (such as `bytes` comparison) while preserving the correct signed token ordering.
> 
> What's the benefit of this approach over just storing the token as a signed integer? If I understand correctly, we already decided that the token will always be a 64-bit integer, and we'll never go back to other possibilities like bytes?

Backwards compatibility.

> 
> > Add new `unsigned_token_column_computation`, which computes using new `unsigned_data()` method.
> > Add new `correct_idx_token_in_secondary_index` feature, which flags that the whole cluster is able to use new `unsigned_token_column_computation`.
> 
> We don't have to do this now, but I wonder if instead of adding a new cluster feature for every new computed-column computation, we could somehow make every computed column computation into a cluster feature? @psarna could that make sense? I'm asking this because we'll probably need such a feature for every computed-column computation class we want to add in the future.
> 
> > Fix tests according to new token ordering
> 
> Sad (it means we had tests which were wrong, and enshrined wrong behavior).
> 
> > and add one new test to validate this aspect explicitly.
> 
> Good. I also wrote a test for this feature in Python which I attached in the issue, which unfortunately I can't get into the repository because @kostja hates shell scripts and I didn't yet have time to convert it to Python.
> 
> > Fixes #7443
> > Tested manually a scenario when someone created an index on old version of Scylla and then migrated to new Scylla. Old index continued to work properly (but returning in wrong order). Upon dropping and re-creating the index, it still returned the same data, but now in correct order.
> 
> Great.

File: bytes.hh
Comment (nyh) [reply bytes.hh:72]: Also, why move this function into bytes.hh? This weird function is only used in token.hh, so I think it should be there. If we ever need it anywhere else (I hope we don't!) we can move it then, and probably will need to add a comment what it does.

File: dht/token.hh
Comment (nyh) [inline dht/token.hh:108]: Is this transformation the same as "bias()" in dht/token.hh? Would have been nice if we had this function once in a header file, with decent comments. But we don't have to do it now.

File: dht/token.hh
Comment (nyh) [inline dht/token.hh:104]: This function needs a comment why it exists. It's not at all obvious what does "unsigned data" mean. To be honest, it's also not clear what data() means - it should have had a comment too...

**Comment by avelanarius** [comment]
> > > Fixes returned rows ordering to proper signed token ordering. Before this change, rows were sorted by token, but using unsigned comparison, meaning that negative tokens appeared after positive tokens.
> > > Add new `unsigned_data()` method to `token`, which returns `token` as `bytes`, but first mapping `token` range from `[-2^63, 2^63-1]` to `[0, 2^64-1]`, which allows it to be compared using unsigned comparsions (such as `bytes` comparison) while preserving the correct signed token ordering.
> > 
> > 
> > What's the benefit of this approach over just storing the token as a signed integer? If I understand correctly, we already decided that the token will always be a 64-bit integer, and we'll never go back to other possibilities like bytes?
> 
> Backwards compatibility.
> 
> > > Add new `unsigned_token_column_computation`, which computes using new `unsigned_data()` method.
> > > Add new `correct_idx_token_in_secondary_index` feature, which flags that the whole cluster is able to use new `unsigned_token_column_computation`.
> > 
> > 
> > We don't have to do this now, but I wonder if instead of adding a new cluster feature for every new computed-column computation, we could somehow make every computed column computation into a cluster feature? @psarna could that make sense? I'm asking this because we'll probably need such a feature for every computed-column computation class we want to add in the future.
> > > Fix tests according to new token ordering
> > 
> > 
> > Sad (it means we had tests which were wrong, and enshrined wrong behavior).
> > > and add one new test to validate this aspect explicitly.
> > 
> > 
> > Good. I also wrote a test for this feature in Python which I attached in the issue, which unfortunately I can't get into the repository because @kostja hates shell scripts and I didn't yet have time to convert it to Python.
> > > Fixes #7443
> > > Tested manually a scenario when someone created an index on old version of Scylla and then migrated to new Scylla. Old index continued to work properly (but returning in wrong order). Upon dropping and re-creating the index, it still returned the same data, but now in correct order.
> > 
> > 
> > Great.

@nyh Upon discussion with @haaawk, we decided that we should consider changing type of `idx_token` to `int64` in indexes - but only in new indexes (the "Backwards compatibility" comment referred to the situation if we wanted to change also types in existing indexes).

File: column_computation.hh
Comment (nyh) [reply column_computation.hh:66]: I agree. But the string name of the computation cannot change, so this will create a somewhat confusing correspondence between the function name and the computation name, so we'll need to make this clear in a comment because nobody will ever remember.

File: test/boost/secondary_index_test.cc
Comment (nyh) [inline test/boost/secondary_index_test.cc:1523]: As a sanity check, please make sure this test *fails* before your patch :-)

File: index/secondary_index_manager.cc
Comment (haaawk) [reply index/secondary_index_manager.cc:149]: Probably not to be addressed in this PR though. It would make it way bigger than necessary.

**Comment by avelanarius** [comment]
@haaawk @psarna @nyh Re-did a part of this PR: now instead of using `bytes` with bias, `idx_token` column type is `long_type` (for new indexes). Therefore some of the comments related to changes to `token.hh`, `bytes.hh` are now not relevant (those files are now not modified). After those changes I redid my manual testing.

File: test/boost/secondary_index_test.cc
Comment (avelanarius) [reply test/boost/secondary_index_test.cc:1523]: Tested it: only adding this new test without any other changes - it fails.

File: index/secondary_index_manager.cc
Comment (avelanarius) [reply index/secondary_index_manager.cc]: Renamed to `legacy_token_column_computation` in new commits.

File: column_computation.hh
Comment (avelanarius) [reply column_computation.hh]: > so we'll need to make this clear in a comment because nobody will ever remember.

Added some comments in `column_computation.hh` describing this situation.

**Review by haaawk — APPROVED** [review]
LGTM

**Review by psarna — APPROVED** [review]

File: column_computation.hh
Comment (nyh) [inline column_computation.hh:63]: I didn't understand this sentence. Only on further reading, I understood you meant that it is not used in new tables.

File: column_computation.hh
Comment (nyh) [inline column_computation.hh:84]: I think it's not clear what the phrase "in schema" means. Also I see below it's actually "token_v2", not "version 2" or "v2". Perhaps you can have a sentence like:
The type name stored for computations of this class is "token_v2" (the name "token" refers to the older legacy_token_column_computation).

**Review by nyh — CHANGES_REQUESTED** [review]
Looks good! I only have a couple of minor requests.

**Comment by avelanarius** [comment]
@nyh Fixed the comments (in amended commits) in accordance with your suggestions.

**Review by nyh — APPROVED** [review]

**Comment by haaawk** [comment]
I think we could merge this now after 3 people has approved this PR, no?

**Comment by nyh** [comment]
Yes, I'm testing to merge it right now.

**Comment by nyh** [comment]
I'm seeing a failure with the test I wrote in python so I'm still debugging if it's a bug in my test or in the implementation. I'll merge this (or not) later tonight.

**Comment by nyh** [comment]
My test a bug which caused it to sometimes fail. After fixing it, the code here seems to pass my test. So I merged this PR. Thanks.

**Comment by haaawk** [comment]
Thanks @nyh

**Pull request closed without merging**
